File: gomsic_core/parsers/usb.py
"""Parser for USB device information from msinfo32.log.

Extracts the USB section from the already-parsed msinfo32 data.
"""

from __future__ import annotations

import logging
import re
from typing import Optional

from ..debug_log.trace import ParserTraceContext
from ..extractor import ArchiveLayout
from ..models import USBDevice, USBInfo
from .base import BaseParser

logger = logging.getLogger(__name__)


class USBParser(BaseParser):
    name = "usb"

    def parse(self, layout: ArchiveLayout, ctx: ParserTraceContext) -> Optional[USBInfo]:
        if layout.gomsic_dir is None:
            ctx.skip("gomsic directory not found")
            return None

        msinfo_path = layout.gomsic_dir / "msinfo32.log"
        ctx.file_searched(str(msinfo_path))

        if not msinfo_path.is_file():
            ctx.skip("msinfo32.log not found")
            return None

        ctx.file_found(str(msinfo_path))

        text = self.read_utf16_file(msinfo_path)
        if text is None:
            ctx.fail("Could not read msinfo32.log")
            return None

        ctx.file_parsed(str(msinfo_path))

        info = USBInfo()

        # Look for USB section in msinfo32.log
        # Section markers vary: [USB], "USB", or lines containing "USB" in headers
        usb_section = self._extract_usb_section(text)
        if usb_section is None:
            ctx.note("No USB section found in msinfo32.log")
            return None

        # Parse USB entries
        for line in usb_section.splitlines():
            line = line.strip()
            if not line:
                continue

            parts = line.split("\t")
            if len(parts) >= 1:
                device = USBDevice(name=parts[0].strip())
                if len(parts) >= 2:
                    device.device_id = parts[1].strip()
                if len(parts) >= 3:
                    device.status = parts[2].strip()
                info.devices.append(device)

        ctx.note(f"Found {len(info.devices)} USB devices")
        return info if info.devices else None

    def _extract_usb_section(self, text: str) -> Optional[str]:
        """Extract the USB section from msinfo32.log text."""
        # Try [USB] section header
        usb_match = re.search(
            r"\[USB\](.*?)(?:\[|\Z)",
            text,
            re.DOTALL | re.IGNORECASE,
        )
        if usb_match:
            return usb_match.group(1)

        # Try looking for "USB" as a section in tab-delimited format
        lines = text.splitlines()
        in_usb = False
        usb_lines = []
        for line in lines:
            if re.match(r"^\s*USB\s*$", line, re.IGNORECASE):
                in_usb = True
                continue
            if in_usb:
                if line.strip() and not line.startswith("\t") and not line.startswith(" "):
                    break  # New section
                usb_lines.append(line)

        return "\n".join(usb_lines) if usb_lines else None

File: pwa/src/worker/parsers/usb.ts
// ============================================================
// parsers/usb.ts — Port of Python USBParser
// Reads <gomsicDir>/msinfo32.log (UTF-16), extracts [USB] section
// ============================================================

import { BaseParser, ParserContext } from './base'
import type { USBInfo, USBDevice } from '../models'

export class USBParser extends BaseParser<USBInfo> {
  readonly name = 'usb'

  parse(ctx: ParserContext): USBInfo | null {
    const { layout, gomsicDir } = ctx
    if (!gomsicDir) return null

    const vpath = `${gomsicDir}/msinfo32.log`
    const bytes = layout.files.get(vpath)
    if (!bytes) return null

    // msinfo32.log is UTF-16 LE
    const text = this.getUtf16(layout, vpath) ?? this.decodeAuto(bytes)
    if (!text) return null

    // Find [USB] section
    const usbSectionMatch = text.match(/\[USB\]([\s\S]*?)(?=\[|$)/i)
    if (!usbSectionMatch) return null

    const usbSection = usbSectionMatch[1]
    const devices: USBDevice[] = []

    for (const line of usbSection.split(/\r?\n/)) {
      const stripped = line.trim()
      if (!stripped) continue

      // Skip header row (contains "Name" and "Device ID" as column labels)
      if (/^name\s*\t/i.test(stripped) || stripped.toLowerCase() === 'name') continue
      if (/^item\s*\t/i.test(stripped)) continue

      // Tab-delimited: name[0], device_id[1], status[2]
      const parts = stripped.split('\t').map((p) => p.trim())
      if (parts.length === 0 || !parts[0]) continue

      const device: USBDevice = {
        name: parts[0],
        device_id: parts[1] ?? undefined,
        status: parts[2] ?? undefined,
      }

      // Skip rows that look like column headers
      if (
        parts[0].toLowerCase() === 'name' ||
        parts[0].toLowerCase() === 'item'
      ) {
        continue
      }

      devices.push(device)
    }

    if (devices.length === 0) return null

    return { devices }
  }
}
